gandalf.js

```diff
@@ -1,109 +1,99 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?